src/Components/Server/src/CircuitDisconnectMiddleware.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Server.Circuits;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Components.Server
+{
+    internal class CircuitDisconnectMiddleware
+    {
+        private const string CircuitIdKey = "circuitId";
+
+        public CircuitDisconnectMiddleware(
+            ILogger<CircuitDisconnectMiddleware> logger,
+            CircuitRegistry registry,
+            CircuitIdFactory circuitIdFactory,
+            RequestDelegate next)
+        {
+            Logger = logger;
+            Registry = registry;
+            CircuitIdFactory = circuitIdFactory;
+            Next = next;
+        }
+
+        public ILogger<CircuitDisconnectMiddleware> Logger { get; }
+        public CircuitRegistry Registry { get; }
+        public CircuitIdFactory CircuitIdFactory { get; }
+        public RequestDelegate Next { get; }
+
+        public async Task Invoke(HttpContext context)
+        {
+            if (!HttpMethods.IsPost(context.Request.Method))
+            {
+                context.Response.StatusCode = StatusCodes.Status405MethodNotAllowed;
+                return;
+            }
+
+            var form = await context.Request.ReadFormAsync();
+            if(!form.TryGetValue(CircuitIdKey, out var circuitId) || !CircuitIdFactory.ValidateCircuitId(circuitId))
+            {
+                context.Response.StatusCode = StatusCodes.Status404NotFound;
+                return;
+            }
+
+            await TerminateCircuitGracefully(circuitId);
+
+            context.Response.StatusCode = StatusCodes.Status200OK;
+            return;
+        }
+
+        private async Task TerminateCircuitGracefully(string circuitId)
+        {
+            try
+            {
+                Log.CircuitTerminatedGracefully(Logger, circuitId);
```

Comment:
Should this be after the call to `Terminate`? The log is going to say that we terminated gracefully, even if it throws.

Reply:
Fair, will move.